src/shared/copy.c

```diff
@@ -638,6 +638,10 @@ static void hardlink_context_destroy(HardlinkContext *c) {
          * _cleanup_() so that we really delete this, even on failure. */
 
         if (c->dir_fd >= 0) {
+                /* <dir_fd> might be have already been used for reading, so we need to rewind it. */
```

Comment:
This is obscure enough that I think it deserves a comment in src/shared/rm-rf.h to warn future users to be careful about this